File: ios/DailyMemory/Domain/UseCase/Memory/SearchMemoriesUseCase.swift
import Foundation

/// Use case for searching memories with various filters.
final class SearchMemoriesUseCase {
    private let memoryRepository: MemoryRepository

    init(memoryRepository: MemoryRepository = MemoryRepositoryImpl()) {
        self.memoryRepository = memoryRepository
    }

    // Search by content
    func byContent(query: String) async throws -> [Memory] {
        return try await memoryRepository.search(query: query)
    }

    // Search by category
    func byCategory(_ category: Category) async throws -> [Memory] {
        return try await memoryRepository.getByCategory(category)
    }

    // Search by date range
    func byDateRange(from: Date, to: Date) async throws -> [Memory] {
        return try await memoryRepository.getByDateRange(from: from, to: to)
    }

    // Search by specific date
    func byDate(_ date: Date) async throws -> [Memory] {
        return try await memoryRepository.getByDate(date)
    }

    // Search by person
    func byPerson(personId: String) async throws -> [Memory] {
        return try await memoryRepository.getByPersonId(personId)
    }
}

File: android/app/src/main/java/com/dailymemory/domain/usecase/memory/SearchMemoriesUseCase.kt
package com.dailymemory.domain.usecase.memory

import com.dailymemory.domain.model.Category
import com.dailymemory.domain.model.Memory
import com.dailymemory.domain.repository.MemoryRepository
import kotlinx.coroutines.flow.Flow
import java.time.LocalDateTime
import javax.inject.Inject

/**
 * Use case for searching memories with various filters.
 */
class SearchMemoriesUseCase @Inject constructor(
    private val memoryRepository: MemoryRepository
) {
    // Main invoke operator for simple search
    operator fun invoke(query: String): Flow<List<Memory>> {
        return memoryRepository.searchByContent(query)
    }

    // Search by content
    fun byContent(query: String): Flow<List<Memory>> {
        return memoryRepository.searchByContent(query)
    }

    // Search by category
    fun byCategory(category: Category): Flow<List<Memory>> {
        return memoryRepository.observeByCategory(category)
    }

    // Search by date range
    fun byDateRange(start: LocalDateTime, end: LocalDateTime): Flow<List<Memory>> {
        return memoryRepository.observeByDateRange(start, end)
    }

    // Search by specific date
    fun byDate(date: LocalDateTime): Flow<List<Memory>> {
        return memoryRepository.observeByDate(date)
    }

    // Search by person
    fun byPerson(personId: String): Flow<List<Memory>> {
        return memoryRepository.observeByPersonId(personId)
    }
}
